prow/pubsub/subscriber/server.go

```diff
@@ -0,0 +1,230 @@
+	}
+	return s.sub.Receive(ctx, g)
+}
+
+func (c *PubSubClient) New(ctx context.Context, project string) (pubsubClientInterface, error) {
+	client, err := pubsub.NewClient(ctx, project)
+	if err != nil {
+		return nil, err
+	}
+	c.client = client
+	return c, nil
+}
+
+func (c *PubSubClient) Subscription(id string) subscriptionInterface {
+	return &pubsubSubscription{
+		sub: c.client.Subscription(id),
+	}
+}
+
+// handlePulls pull for Pub/Sub subscriptions and handle them.
+func (s *PullServer) handlePulls(ctx context.Context, projectSubscriptions config.PubsubSubscriptions) (*errgroup.Group, context.Context, error) {
+	// Since config might change we need be able to cancel the current run
+	errGroup, derivedCtx := errgroup.WithContext(ctx)
+	for project, subscriptions := range projectSubscriptions {
+		client, err := s.Client.New(ctx, project)
+		if err != nil {
+			return errGroup, derivedCtx, err
+		}
+		for _, subName := range subscriptions {
+			sub := client.Subscription(subName)
+			errGroup.Go(func() error {
+				logrus.Infof("Listening for subscription %s on project %s", sub.String(), project)
+				defer logrus.Warnf("Stopped Listening for subscription %s on project %s", sub.String(), project)
+				err := sub.Receive(derivedCtx, func(ctx context.Context, msg messageInterface) {
+					if err = s.Subscriber.handleMessage(msg, sub.String()); err != nil {
+						s.Subscriber.Metrics.ACKMessageCounter.With(prometheus.Labels{subscriptionLabel: sub.String()}).Inc()
+					} else {
+						s.Subscriber.Metrics.NACKMessageCounter.With(prometheus.Labels{subscriptionLabel: sub.String()}).Inc()
+					}
+					msg.Ack()
+				})
+				if err != nil {
+					return err
+				}
+				return nil
+			})
+		}
+	}
+	return errGroup, derivedCtx, nil
+}
+
+// Run will block listening to all subscriptions and return once the context is cancelled
+// or one of the subscription has a unrecoverable error.
+func (s *PullServer) Run(ctx context.Context) error {
+	var err error
+	defer func() {
+		if err != nil {
+			logrus.WithError(ctx.Err()).Error("Pull server shutting down")
+		}
+		logrus.Warn("Pull server shutting down")
+	}()
+	currentConfig := s.Subscriber.ConfigAgent.Config().PubsubSubscriptions
+	errGroup, derivedCtx, err := s.handlePulls(ctx, currentConfig)
+	if err != nil {
+		return err
+	}
+
+	for {
+		select {
+		// Parent context. Shutdown
+		case <-ctx.Done():
+			return ctx.Err()
+		// Current thread context, it may be failing already
+		case <-derivedCtx.Done():
+			err = errGroup.Wait()
+			return err
+		// Checking for update config
+		case <-s.ConfigCheckTick.C:
+			newConfig := s.Subscriber.ConfigAgent.Config().PubsubSubscriptions
+			logrus.Info("Checking for new config")
+			if !reflect.DeepEqual(newConfig, currentConfig) {
+				logrus.Warn("New config found, reloading pull Server")
+				// Making sure the current thread finishes before starting a new one.
+				errGroup.Wait()
+				// Starting a new thread with new config
```

Comment:
What are the worker in your idea ? handlePulls ? By the way there is only one worker.
We still need to check if the config changed, and it is just easier to do it in the parent, such that we can cancel the child if a new config exist. Also I prefer passing the config down because from the parent since the parent is already doing a lookup. If the child is doing another lookup the parent and the child might not have the same config, and therefore the parent might issue an extra cancel to the child.

I considered keeping track of all the sub.Receive threads, and removing / adding them when the config is changed, but that s much more complex and prone to more bugs. I could not see the benefits of doing this over just restarting all the threads when a new config is here since pubsub will resend anyway if a message is not acked.

Reply:
Hmmm. True, there is only one `handlePulls`. In one of my WIP pulls I have a config change subscription -- let me post that and we could use one impl.

Reply:
I was thinking something like this https://github.com/kubernetes/test-infra/pull/10247